rqd/rqd/rqnimby.py

```diff
@@ -64,114 +76,210 @@ def lockNimby(self):
+
+class NimbyPynput(Nimby):
+    def __init__(self, rqCore):
+        Nimby.__init__(self, rqCore)
+
+        import pynput
+        self.mouse_listener = pynput.mouse.Listener(
+            on_move=self.on_interaction,
+            on_click=self.on_interaction,
+            on_scroll=self.on_interaction)
+        self.keyboard_listener = pynput.keyboard.Listener(on_press=self.on_interaction)
+
+    def on_interaction(self, *args):
+        self.interaction_detected = True
+
+    def startListener(self):
+        self.mouse_listener.start()
+        self.keyboard_listener.start()
+
+    def stopListener(self):
+        self.mouse_listener.stop()
+        self.keyboard_listener.stop()
+
+    def lockedInUse(self):
+        self.interaction_detected = False
+
+        time.sleep(5)
+        if self.active and self.interaction_detected == False:
+            self.lockedIdle()
+        elif self.active:
+
             self.thread = threading.Timer(rqd.rqconstants.CHECK_INTERVAL_LOCKED,
                                           self.lockedInUse)
             self.thread.start()
 
     def unlockedIdle(self):
-        """Nimby State: Machine is idle, host is unlocked,
-                        waiting for user activity"""
         while self.active and \
-              self.results[0] == [] and \
-              self.rqCore.machine.isNimbySafeToRunJobs():
-            try:
-                self._openEvents()
-                self.results = select.select(self.fileObjList, [], [], 5)
-            except:
-                pass
+                self.interaction_detected == False and \
+                self.rqCore.machine.isNimbySafeToRunJobs():
+
+            time.sleep(5)
+
             if not self.rqCore.machine.isNimbySafeToRunJobs():
                 log.warning("memory threshold has been exceeded, locking nimby")
                 self.active = True
 
         if self.active:
-            self._closeEvents()
             self.lockNimby()
             self.thread = threading.Timer(rqd.rqconstants.CHECK_INTERVAL_LOCKED,
                                           self.lockedInUse)
             self.thread.start()
 
-    def run(self):
-        """Starts the Nimby thread"""
-        self.active = True
-        self.unlockedIdle()
+    def lockedIdle(self):
+        waitStartTime = time.time()
 
-    def stop(self):
-        """Stops the Nimby thread"""
-        if self.thread:
-            self.thread.cancel()
-        self.active = False
-        self._closeEvents()
-        self.unlockNimby()
+        time.sleep(rqd.rqconstants.MINIMUM_IDLE)
+
+        if self.active and self.interaction_detected == False and \
+                self.rqCore.machine.isNimbySafeToUnlock():
+
+            self.unlockNimby(asOf=waitStartTime)
+            self.unlockedIdle()
+        elif self.active:
+
+            self.thread = threading.Timer(rqd.rqconstants.CHECK_INTERVAL_LOCKED,
+                                          self.lockedInUse)
+            self.thread.start()
```

Comment:
nit: Let's leave a blank line at the end of every file. Usually an IDE setting you can use to do that on save.

Reply:
Done